source/op/descrpt_se_a_mask.cc

```diff
@@ -0,0 +1,337 @@
+
+#include "custom_op.h"
+#include "ComputeDescriptor.h"
+#include "neighbor_list.h"
+#include "fmt_nlist.h"
+#include "errors.h"
+#include <cmath>
+
+typedef double boxtensor_t;
+typedef double compute_t;
+
+REGISTER_OP("DescrptSeAMask")
+    .Attr("T: {float, double} = DT_DOUBLE")
+    .Input("coord: T")
+    .Input("type: int32")
+    .Input("mask: int32")
+    .Attr("total_atom_num: int")
+    .Output("descrpt: T")
+    .Output("descrpt_deriv: T")
+    .Output("rij: T")
+    .Output("nlist: int32");
+
+template <typename Device, typename FPTYPE>
+class DescrptSeAMaskOp : public OpKernel
+{
+public:
+    explicit DescrptSeAMaskOp(OpKernelConstruction *context) : OpKernel(context)
+    {
+        OP_REQUIRES_OK(context, context->GetAttr("total_atom_num", &total_atom_num));
+    }
+
+    void Compute(OpKernelContext *context) override
+    {
+        deepmd::safe_compute(context, [this](OpKernelContext *context)
+                             { this->_Compute(context); });
+    }
+
+    void _Compute(OpKernelContext *context)
+    {
+        // Grab the input tensor
+        int context_input_index = 0;
+        const Tensor &coord_tensor = context->input(context_input_index++);
+        const Tensor &type_tensor = context->input(context_input_index++);
+        const Tensor &mask_matrix_tensor = context->input(context_input_index++);
+
+        // set size of the sample
+        OP_REQUIRES(context, (coord_tensor.shape().dims() == 2), errors::InvalidArgument("Dim of coord should be 2"));
+        OP_REQUIRES(context, (type_tensor.shape().dims() == 2), errors::InvalidArgument("Dim of type for se_e2_a_mask op should be 2"));
+        OP_REQUIRES(context, (mask_matrix_tensor.shape().dims() == 2), errors::InvalidArgument("Dim of mask matrix should be 2"));
+
+        int nsamples = coord_tensor.shape().dim_size(0);
+
+        // check the sizes
+        OP_REQUIRES(context, (nsamples == type_tensor.shape().dim_size(0)), errors::InvalidArgument("number of samples should match"));
+        OP_REQUIRES(context, (nsamples == mask_matrix_tensor.shape().dim_size(0)), errors::InvalidArgument("number of samples should match"));
+
+        // Set n_descrpt for each atom. Include 1/rr, cos(theta), cos(phi), sin(phi)
+        int n_descrpt = 4;
+
+        // Create an output tensor
+        TensorShape descrpt_shape;
+        descrpt_shape.AddDim(nsamples);
+        descrpt_shape.AddDim(total_atom_num * total_atom_num * n_descrpt);
```

Comment:
This may need to be cast to `int64` for huge systems.

Reply:
In my thought, mask model will be applied small system, it can't afford a >200 particles system. Thus, int32 might be enough for `total_atom_num * total_atom_num * n_descrpt`.